mcu/atmega-hal/src/adc.rs

```diff
@@ -1,56 +1,8 @@
 //! Analog-to-Digital Converter
 
-use crate::port;
+use crate::{pac, port::*};
 pub use avr_hal_generic::adc::{AdcChannel, AdcOps, ClockDivider};
 
-/// Select the voltage reference for the ADC peripheral
-///
-/// The internal voltage reference options may not be used if an external reference voltage is
-/// being applied to the AREF pin.
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
-#[repr(u8)]
-pub enum ReferenceVoltage {
-    /// Voltage applied to AREF pin.
-    Aref,
-    /// Default reference voltage (default).
-    AVcc,
-    /// Internal reference voltage.
-    Internal,
-}
-
-impl Default for ReferenceVoltage {
-    fn default() -> Self {
-        Self::AVcc
-    }
-}
-
-/// Configuration for the ADC peripheral.
-#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
-pub struct AdcSettings {
-    pub clock_divider: ClockDivider,
-    pub ref_voltage: ReferenceVoltage,
-}
-
-fn apply_settings(peripheral: &crate::pac::ADC, settings: AdcSettings) {
-    peripheral.adcsra.write(|w| {
-        w.aden().set_bit();
-        match settings.clock_divider {
-            ClockDivider::Factor2 => w.adps().prescaler_2(),
-            ClockDivider::Factor4 => w.adps().prescaler_4(),
-            ClockDivider::Factor8 => w.adps().prescaler_8(),
-            ClockDivider::Factor16 => w.adps().prescaler_16(),
-            ClockDivider::Factor32 => w.adps().prescaler_32(),
-            ClockDivider::Factor64 => w.adps().prescaler_64(),
-            ClockDivider::Factor128 => w.adps().prescaler_128(),
-        }
-    });
-    peripheral.admux.write(|w| match settings.ref_voltage {
-        ReferenceVoltage::Aref => w.refs().aref(),
-        ReferenceVoltage::AVcc => w.refs().avcc(),
-        ReferenceVoltage::Internal => w.refs().internal(),
-    });
-}
```

Comment:
I'm unsure about moving this block into the macro when it was easily possible to have the code sit in `atmega-hal` without unnecessary duplication.

I would only push code into `avr-hal-generic` when there are multiple HAL crates that need the same implementation.  Any code that can stay in `atmega-hal` makes things stay more clear and also doesn't contribute to bloat of the `avr-hal-generic`.

Reply:
(As an example of a case where I very much agree with what you did is the `set_channel` impl:  I'm glad to see it deduplicated)

Reply:
Yeah, I moved this early on with the intent to generate the `set_reference` and `set_channel` blocks, for which I'd need to know the variants. But later I discovered this was more troublesome than I anticipated, since the attinys have the `refs2` bit separate, and ended up leaving the block to the macro's caller. I'll check if I can refactor this to generate these blocks, else move them back. It really is redundant as it is, the macro gets the variants and just passes them along untouched.